### PR TITLE
CA-219257: Make enic driver workaround configurable.

### PR DESCRIPTION
**Pull request by sharady**
From xcp-networkd.conf file we can configure enic-workaround-until-version="x.x.x.x"
to the version till enic driver workaround will be applied or
set the version to an empty string for not applying the workaround.

Signed-off-by: Sharad Yadav <sharad.yadav@citrix.com>

### CODE REVIEW

File: lib/network_utils.ml [hunk 1]
```diff
@@ -74,6 +74,13 @@ module Sysfs = struct
 			warn "Failed to obtain list of drivers from sysfs";
 			[]
 
+	let get_driver_version driver () =
+		try
+			String.strip String.isspace (Unixext.string_of_file ("/sys/bus/pci/drivers/" ^ driver ^ "/module/version"))
+		with _ ->
+			warn "Failed to obtain driver version from sysfs";
+			""
```
Comment (robhoes) [inline lib/network_utils.ml]: Using an option type would be better.

**Comment by lindig** [comment]
I don't have a string opinion about this but would use a different name for the flag: `enic_workaround_enabled=true/false`. It is shorter and shares with the other flag the same `enic` prefix.

**Comment by robhoes** [comment]
The `enic_driver_version` config option should have a name that more clearly indicates what it is for, e.g. `enic-workaround-until-version`. We can also combine the two options into one, and say that the workaround is disabled if `enic-workaround-until-version` is the empty string.

File: networkd/network_server.ml [hunk 2]
```diff
@@ -85,12 +87,22 @@ let set_dns_interface _ dbg ~name =
 	debug "Setting DNS interface to %s" name;
 	config := {!config with dns_interface = Some name}
 
+let compare_driver_version vs1 vs2 () =
+	try
+		let version_of vs = List.map int_of_string (String.split '.' vs) in
+		let first_non_zero lst = List.fold_left (fun a b -> if (a<>0) then a else b) 0 lst in
+		(first_non_zero (List.map2 compare (version_of vs1) (version_of vs2)) <= 0)
```
Comment (robhoes) [inline networkd/network_server.ml]: This may not do the right thing if the lists are not the same size.

Comment (robhoes) [inline networkd/network_server.ml]: I think this function is a bit hard to follow, and it took me a while to figure out how it is meant to work. Also, it is not obvious what the return value of the function actually means. Also, it looks like `version_of` is really more like a `list_of_version`...

How about something like this:
```
(* Returns `true` is vs1 is older than vs2 *)
let is_older_version vs1 vs2=
  let list_of_version vs = List.map int_of_string (String.split '.' vs) in
  let rec loop vs1' vs2' =
    match vs1', vs2' with
    | [], _ | _, [] -> false
    | a :: _, b :: _ when a < b -> true
    | _ :: tl1, _ :: tl2 -> loop tl1 tl2
  in
  loop (list_of_version vs1) (list_of_version vs2)
```

Comment (sharady) [inline networkd/network_server.ml]: This code looks more relaxed for comparing driver version in case version string having different count of digits. Thanks @robhoes I have updated the PR as per review comments.
